posts/the-game-of-vue-migrations.md

```diff
@@ -0,0 +1,157 @@
+
+Removing the useless `example.spec.js` and `HelloWorld.vue` files and adding a basic test for `App.js`, the output is now much more helpful:
+
+![Unit Testing Output with Code Coverage](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/VueUnitTestingOutputCodeCoverage.png)
+
+One last tip before bringing up the code coverage, add this command to the `"scripts"` section of the `package.json` to run the tests continuously:
+
+```json
+"test:unit:watch": "vue-cli-service test:unit --watch --coverage"
+```
+
+### Test-Driven Development vs. Testing as an Afterthought
+
+When testing is at the core of an application's development cycle, this is known as [Test-Driven Development (TDD)](https://en.wikipedia.org/wiki/Test-driven_development). Writing tests first is a different mindset than writing tests after the application code. Writing tests before changing the codebase requires exceptional discipline. Unless it is well-practiced, testing will always be more an afterthought than a designing methodology. The good news is once there are written tests, then the shift in mindset can begin to occur, and the power of TDD can begin to take hold.
+
+With enough tests to bring the coverage reports up to compliant levels (generally think >80% but be sure to hit any edge cases!), the console output may look more like this:
+
+![Unit Testing Output with Full Code Coverage](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/VueUnitTestingOutputFullCodeCoverage.png)
+
+Thinking ahead while writing the tests to the refactoring/TypeScript step, here are a few opportunities for improvement:
+
+1. The default "`grid`" value of "`[[false]]`" within "`data()`" is an incorrect placeholder.
+1. There are variable name typos, inconsistent function names, and [confusing function responsibilities](https://en.wikipedia.org/wiki/Single-responsibility_principle).
+1. The "`checkNeighbors`" method requires more intricate tests. Due to this, and as the [SonarLint extension](https://www.sonarlint.org/vscode) suggests, the method's complexity needs reducing.
+
+![CheckNeighbors Method Complexity](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2021/10/VueCheckNeighborsComplexity.png)
+
+So what is the next step? Is it to refactor/add TypeScript or upgrade to Vue 3? Considering that Vue 3 is natively in TypeScript, performing the transition to TypeScript first is probably best. That way, an upgrade to Vue 3 would be even more straightforward. It could allow for more than just following [the above guidance](https://v3.vuejs.org/guide/migration/migration-build.html), like generating a new Vue 3 project then copying over the existing Vue components.
```

Comment:
```suggestion
So, what is the next step? Is it to refactor/add TypeScript or upgrade to Vue 3? Considering that Vue 3 is natively in TypeScript, performing the transition to TypeScript first is best. That way, an upgrade to Vue 3 would be even more straightforward. It could allow for more than just following [the above guidance](https://v3.vuejs.org/guide/migration/migration-build.html), like generating a new Vue 3 project then copying over the existing Vue components.
```